src/topology/homotopy/contractible.lean

```diff
@@ -0,0 +1,79 @@
+/-
+Copyright (c) 2022 Praneeth Kolichala. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Praneeth Kolichala
+-/
+
+import topology.homotopy.path
+import topology.homotopy.equiv
+
+/-!
+# Contractible spaces
+
+In this file, we define `contractible_space`, a space that is homotopy equivalent to `unit`.
+-/
+
+noncomputable theory
+
+namespace continuous_map
+variables {X Y Z : Type*} [topological_space X] [topological_space Y] [topological_space Z]
+
+/-- A map is nullhomotopic if it is homotopic to a constant map. -/
+def nullhomotopic (f : C(X, Y)) : Prop :=
+∃ y : Y, homotopic f (continuous_map.const _ y)
+
+lemma nullhomotopic_of_constant (y : Y) : nullhomotopic (continuous_map.const X y) :=
+⟨y, by refl⟩
+
+lemma nullhomotopic_compose_right {f : C(X, Y)} (hf : f.nullhomotopic) (g : C(Y, Z)) :
+  (g.comp f).nullhomotopic :=
+by { cases hf with y hy, use g y, exact homotopic.hcomp hy (homotopic.refl g), }
+
+lemma nullhomotopic_compose_left {f : C(Y, Z)} (hf : f.nullhomotopic) (g : C(X, Y)) :
+  (f.comp g).nullhomotopic :=
+by { cases hf with y hy, use y, exact homotopic.hcomp (homotopic.refl g) hy, }
+
+end continuous_map
+
+open continuous_map
+open_locale continuous_map
+
+/-- A contractible space is one that is homotopy equivalent to `unit`. -/
+class contractible_space (X : Type*) [topological_space X] : Prop :=
+(hequiv_unit [] : nonempty (X ≃ₕ unit))
+
+
+variables (X : Type*) [topological_space X] [contractible_space X]
+
+lemma id_nullhomotopic : (continuous_map.id X).nullhomotopic :=
+begin
+  obtain ⟨hv⟩ := contractible_space.hequiv_unit X,
+  use hv.inv_fun (),
+  convert hv.left_inv.symm,
+  ext, simp, congr,
+end
+
+lemma contractible_iff_id_nullhomotopic (Y : Type*) [topological_space Y] :
+  contractible_space Y ↔ (continuous_map.id Y).nullhomotopic :=
+begin
+  split, { introI, apply id_nullhomotopic, },
+  rintro ⟨p, h⟩,
+  refine_struct { hequiv_unit := ⟨{ to_fun := continuous_map.const _ (),
+    inv_fun := continuous_map.const _ p }⟩ },
+  { exact h.symm, }, { convert homotopic.refl (continuous_map.id unit), ext, },
+end
+
+namespace contractible_space
+
+@[priority 50]
+instance : nonempty X := nonempty.map (λ h, homotopy_equiv.inv_fun h ()) (hequiv_unit X)
```

Comment:
I'm a little worried about making this an instance, even at low priority. `nonempty` is used in a lot of places where topology is irrelevant. I'd be perfectly happy if this were just a `def`, or we can get a second opinion on it being an instance.

Reply:
I changed it to `lemma` (which is what I assume you meant since `nonempty` is `Prop`-valued), but I don't know exactly how the instance search algorithm works at a lower level, so I don't know how much of an issue/non-issue this really is.

Reply:
Actually, I realized `path_connected` has a `nonempty` instance on it already. So this instance was completely redundant. I'm going to probably delete the lemma altogether.